Re-Enable Category filtering in WP list

NobodysNightmare opened the pull request:
## OpenProject work package

https://community.openproject.org/work_packages/19759
public WP: https://community.openproject.org/work_packages/19809
## Description

This PR adds category filters to the WP list. Those filters were available before and simply forgotten to implement.

For better forwards compatibility I did not invent an experimental API endpoint for categories, but used the one introduced in release 4.1. That involved some backporting, see the notes below.
## :warning: :warning: :warning:  Important notes :warning: :warning: :warning:

Commit https://github.com/opf/openproject/commit/23e7b113c4af39f76d1cd733aeb808521c0a1806 contains lots of code that was backported from `release/4.1`.

You should merge that **commit** (read: not the PR, but the sole commit) into the `release/4.1` branch using the `ours` strategy:

```
git checkout release/4.1
git merge -s ours 23e7b113c4af39f76d1cd733aeb808521c0a1806 -m "Resolving forward-merge of backport commit"
```

The afforementioned commit contains things like ridicuously simplified specs (to have basic specs, but to not let them fail or backport them in a 4.0 compatible manner; e.g. error cases).
It should be considered poisonous for everything else, but the `release/4.0` branch.

The rest of this PR can be merged safely and should not contain unhealthy ingredients.

NobodysNightmare committed:
Backporting APIv3 categories to OP 4.0

Backport was neccessary to implement missing categories filter without adding more technical debt.
These changes will be merged into 4.1 and following branches and vanish in that merge (merge-strategy "ours")

YOU SHOULD NOT SEE ANY CONTENTS FROM THIS COMMIT ANYWHERE AFTER 4.0.x!

This commit was built from the commits of the GitHub-PR mentioned below, as well as some code dependencies for that.

The GitHub-PR was https://github.com/opf/openproject/pull/2505

NobodysNightmare committed:
WIP wire up category filters in frontend

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-fix-category-filter.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-fix-category-filter.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

Commit:
parse categories from api response

finish up the groundwork laid out by @NobodysNightmare and actually parse the
response fromthe backported api

floriank commented:
This should work now - I have decided to to the porting to `release/4.1` in a separate PR which I'll reference here.

floriank commented:
The port for this to `release/4.1` is open for discussion at #2911

floriank commented:
Note: Public WP - https://community.openproject.org/work_packages/19809

floriank commented:
@NobodysNightmare there is still some sh.. I mean stuff breaking in the postgresql test suite :(

NobodysNightmare commented:
I did what always works: I restarted the offending test job. The issues don't make any sense to me, so before trying to understand them, I try to forget them...

NobodysNightmare commented:
Ha! It worked!
